walk/walk.go

```diff
@@ -356,3 +359,63 @@ func resolveFileInfo(wc *walkConfig, dir, rel string, ent fs.DirEntry) fs.DirEnt
 	}
 	return fs.FileInfoToDirEntry(fi)
 }
+
+type pathTrie struct {
+	children map[string]*pathTrie
+	entry    *fs.DirEntry
+}
+
+// Basic factory method to ensure the entry is properly copied
+func newTrie(entry fs.DirEntry) *pathTrie {
+	return &pathTrie{entry: &entry}
+}
+
+func buildTrie(c *config.Config, isIgnored isIgnoredFunc) (*pathTrie, error) {
+	trie := &pathTrie{
+		children: map[string]*pathTrie{},
+	}
+
+	// A channel to limit the number of concurrent goroutines
+	limitCh := make(chan struct{}, 100)
```

Comment:
`errgroup.SetLimit` means it will fail if you reach that limit (then you can implement a fallback but the PR gets uglier), where using a channel this way means it blocks at 100.

I figure it's always good to set a limit. Initially I was thinking of just using the error group with no limit but found that increased the "system %" (when using `time`) without any gain on the actual total time.

So there's multiple options and it's debatable what's best. I assume something like [what golang/tools does](https://github.com/golang/tools/blob/v0.25.0/internal/gopathwalk/walk.go#L285) is probably best but involves more code or pulling in a dependency such as [fastwalk](https://github.com/charlievieth/fastwalk) which does similar (see [1820](https://github.com/bazelbuild/bazel-gazelle/pull/1820) which uses fastwalk).

Reply:
FWIW I think after this initial commit we can push more computation into the parallel walk, adding more cpu usage (not just io) into here and then this limit might be more important.

Things like parsing the BUILDs doesn't have to happen in walk order etc.

Reply:
Are you sure it fails? From the docs:
```
Any subsequent call to the Go method will block until it can add an active goroutine without exceeding the configured limit.
```
That sounds like exactly what we are doing manually with the channel.

Reply:
Maybe (most likely) I'm misunderstanding something, but when I set I did `SetLimit(100)` I occasionally had deadlock issues. I only saw it once in this repo but saw it consistently in a significantly larger repo.

This is from running `bazel test //...` in `tests/bcr/go_mod`:
```
ERROR: An error occurred during the fetch of repository 'gazelle~~go_deps~com_github_envoyproxy_protoc_gen_validate':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_jason/1c364caf8aebd7c2af742cd83fd3e40a/external/gazelle~/internal/go_repository.bzl", line 353, column 17, in _go_repository_impl
		fail("failed to generate BUILD files for %s: %s" % (
Error in fail: failed to generate BUILD files for github.com/envoyproxy/protoc-gen-validate: fatal error: all goroutines are asleep - deadlock!

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0x140001b6ef8?)
	GOROOT/src/runtime/sema.go:62 +0x2c
sync.(*WaitGroup).Wait(0x1400024ee88)
	GOROOT/src/sync/waitgroup.go:116 +0x74
github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup.(*Group).Wait(0x1400024ee80)
	src/github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup/errgroup.go:56 +0x2c
github.com/bazelbuild/bazel-gazelle/walk.buildTrie(0x140000e8180, 0x100b9c438)
	src/github.com/bazelbuild/bazel-gazelle/walk/walk.go:402 +0xf8
github.com/bazelbuild/bazel-gazelle/walk.Walk(0x140000e8180, {0x140000b01e0?, 0x6, 0x6}, {0x1400009ecd0, 0x1, 0x1}, 0x6?, 0x0?)
	src/github.com/bazelbuild/bazel-gazelle/walk/walk.go:127 +0x194
main.runFixUpdate({0x140000b45a0, 0x82}, 0x16f6b3777?, {0x140000b6010, 0x12, 0x13})
	src/github.com/bazelbuild/bazel-gazelle/cmd/gazelle/fix-update.go:317 +0xa70
main.run({0x140000b45a0?, 0x82?}, {0x140000b6010?, 0x12?, 0x13?})
	src/github.com/bazelbuild/bazel-gazelle/cmd/gazelle/main.go:95 +0x20c
main.main()
	src/github.com/bazelbuild/bazel-gazelle/cmd/gazelle/main.go:72 +0x104

goroutine 18 [chan send]:
github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup.(*Group).Go(0x1400024ee80, 0x1400024f740)
	src/github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup/errgroup.go:71 +0x34
github.com/bazelbuild/bazel-gazelle/walk.walkDir({0x140000b4630, 0x82}, {0x0, 0x0}, 0x1400024ee80, 0x100b9c438, 0x1400029a2f0)
	src/github.com/bazelbuild/bazel-gazelle/walk/walk.go:431 +0x3a8
github.com/bazelbuild/bazel-gazelle/walk.buildTrie.func1()
	src/github.com/bazelbuild/bazel-gazelle/walk/walk.go:399 +0x3c
github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1()
	src/github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup/errgroup.go:78 +0x58
created by github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
	src/github.com/bazelbuild/bazel-gazelle/vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x98
```

Reply:
Adding some color since Jason and I analyzed this together... `SetLimit` on errgroup sets the size of the semaphore which governs how many items can be enqueued, whereas the channel/semaphore in this PR governs how many directories are being processed. The reason this matters is that you can end up in a situation where the errgroup limit is reached and therefore none of the in-flight executions are able to enqueue child dir traversals. They end up blocked on `errgroup.Go` and can't exit, thus leading to deadlock.

Limiting the active executions instead of the queue size (like this PR does) avoids that.

Reply:
That makes sense, thanks for looking into it. I guess this also means that the current scheme can't deadlock as it only blocks at a single point, so some (100) goroutines can always make progress.